test/test_broker.hpp

```diff
@@ -900,6 +1096,47 @@ class test_broker {
                 }
             }
         }
+
+        // Re-activate any saved "shared-subscriptions" that match the newly subscribed topics.
```

Comment:
I don't understand what re-activate mean here.
Here is inside of the function subscribe_handler(). If re-activation means moving shared subscription information from offline to online, then the process should be done at the function connect_handler().
It seems that you already done it at the connect_handler().
So, I guess that the process here should newly add or overwrite existing online shared subscription, and no process required for offline shared subscriptions.

Reply:
If there are shared subscriptions for sessions that are disconnected, but not removed (session lifetime timer) then when a new shared subscription is subscribed then the messages for the saved shared subscription need to be sent to the new subscriber.

Reply:
Note that this is not for newly connected sessions. Its for existing sessions adding a *new* shared subscription to a topic that has shared subscriptions for sessions that are disconnected

Reply:
Hmm... I don't understand yet.
Could you write example sequence ?

1. Client c1 connected to broker with Session Expiry Interval: infinity.
2. Client c1 subscribed t1 with share1 (QoS0)
3. Client c1 disconnected.
4. Client c1 connected to broker with Session Expiry Interval: infinity and Clean Start: false.
   - I said that the stored (offline) shared subscription information should be moved to online here. And if stored publish message that is corresponding to the shared subscriptions, send it to c1 here.
5. Client c1 subscribed t1 with share1 (QoS1)
   - Simply updated QoS0 to QoS1 on c1 t1 share1 subscription. If new QoS is QoS0 then do nothing.

I understand the spec as above.

Reply:
I guess that you misunderstand shared subscriptions spec.
If the algorithm is simple round robin, then the following behavior is expected.
Simple round robin means online session and offline session have the same priority.
Implementer can use online prioritized policy but it is different topic.

1. Client c1 connected to broker with Session Expiry Interval: infinity.
1. Client c2 connected to broker with Session Expiry Interval: infinity.
1. Client c1 subscribed t1 with share1 (QoS0)
1. Client c2 subscribed t1 with share1 (QoS0)
1. Client c1 disconnected.
1. Client c3 publish t1, then the message (a) is delivered to c1 (online session).
1. Client c3 publish t1, then the message (b) is delivered to c2 (offline session).
1. Client c3 publish t1, then the message (c) is delivered to c1 (online session).
1. Client c3 publish t1, then the message (d) is delivered to c2 (offline session).
1. Client c1 connected to broker with Session Expiry Interval: infinity and Clean Start: false.
1. The broker restore c1's shared subscription (it is similar to non shared subscriptions), and the broker deliver the message (a) and (c) to c1.
1. Client c1 subscribed t1 with share1 (QoS0), the broker do nothing.

It is very similar to non shared subscriptions behavior. But deliver to only one of client that have the same share name instead of all clients.

So, restore shared subscriptions and deliver stored message process should be step 10 not step 12